icu4j/demos/src/com/ibm/icu/dev/demo/rbnf/RbnfDemo.java

```diff
@@ -116,8 +116,9 @@ public void textValueChanged(TextEvent e) {
                     textField.setText("PARSE ERROR");
                 }
                 else {
-                    theNumber = new BigDecimal(fieldText);
-                    textField.setText(spelloutFormatter.format(theNumber.doubleValue(), ruleSetName));
+                    theNumber = new BigDecimal(temp instanceof Long ? temp.longValue() : temp.doubleValue());
+                    textField.setText(spelloutFormatter.format(
+                            theNumber.scale() == 0 ? theNumber.longValue() : theNumber.doubleValue(), ruleSetName));
```

Comment:
Why'd you need to change this part?  Does the `BigDecimal` constructor fail to set its scale in such a way that you can tell the difference between an integer and a rational number later on?  Or is there some other special behavior you're getting out of `numberFormatter.parse()` that you don't get out of `BigDecimal`'s constructor?  On the surface, it seems like the code you had before would do what you need and would be more efficient than what you're doing now.

Reply:
Sure, I've changed this part because the input text accepts non-number characters (e.g. letters). So in case the user enters something like "1000abc", if we use the text content directly we can observe exceptions.

So I've used the temp formatted number, which ignores non-number info, and combined with Hoten's suggestion, to guarantee the accuracy and avoid undesired exceptions.

Note: I also though in an improvement by filtering the input type to accept only numbers and decimal separators, but as it's not the focus of the issue, so I just consider the inaccurate aspect.

Reply:
Okay, that sounds reasonable to me.  Thanks for the explanation!